File: backend/routes/notifications.js
const express = require('express');
const router = express.Router();
const Notification = require('../models/Notification');
const { verifyToken } = require('../middleware/auth');

// GET /api/notifications — Get user's notifications
router.get('/', verifyToken, async (req, res) => {
  try {
    const { page = 1, limit = 20 } = req.query;
    const notifications = await Notification.find({ userId: req.user._id })
      .sort({ createdAt: -1 })
      .limit(Number(limit))
      .skip((Number(page) - 1) * Number(limit));

    const unreadCount = await Notification.countDocuments({ userId: req.user._id, read: false });

    res.json({ notifications, unreadCount });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// PATCH /api/notifications/:id/read — Mark as read
router.patch('/:id/read', verifyToken, async (req, res) => {
  try {
    const notif = await Notification.findOneAndUpdate(
      { _id: req.params.id, userId: req.user._id },
      { read: true },
      { new: true }
    );
    if (!notif) return res.status(404).json({ message: 'Notification not found' });
    res.json(notif);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// PATCH /api/notifications/read-all — Mark all as read
router.patch('/read-all', verifyToken, async (req, res) => {
  try {
    await Notification.updateMany({ userId: req.user._id, read: false }, { read: true });
    res.json({ message: 'All notifications marked as read' });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;

File: backend/routers/notifications.py
from fastapi import APIRouter, Depends
from config import db
from dependencies import get_current_user
from bson import ObjectId

router = APIRouter()

@router.get("")
async def get_notifications(user=Depends(get_current_user), limit: int = 50):
    cursor = db.notifications.find({"userId": user["_id"]}).sort("createdAt", -1).limit(limit)
    notifs = []
    async for doc in cursor:
        doc["_id"] = str(doc["_id"])
        doc["userId"] = str(doc["userId"])
        if doc.get("relatedMatch"):
            doc["relatedMatch"] = str(doc["relatedMatch"])
        notifs.append(doc)
    
    unread_count = await db.notifications.count_documents({"userId": user["_id"], "read": False})
    return {"notifications": notifs, "unreadCount": unread_count}

@router.patch("/{id}/read")
async def mark_read(id: str, user=Depends(get_current_user)):
    await db.notifications.update_one(
        {"_id": ObjectId(id), "userId": user["_id"]},
        {"$set": {"read": True}}
    )
    return {"message": "Marked as read"}

@router.patch("/read-all")
async def mark_all_read(user=Depends(get_current_user)):
    await db.notifications.update_many(
        {"userId": user["_id"], "read": False},
        {"$set": {"read": True}}
    )
    return {"message": "All marked as read"}
